src/engines/dropped-frames-watcher.js

```diff
@@ -0,0 +1,69 @@
+// @flow
+import FakeEventTarget from '../event/fake-event-target';
+import getLogger from '../utils/logger';
+
+class DroppedFramesWatcher extends FakeEventTarget {
+  _droppedFramesInterval: ?number = null;
+  _lastDroppedFrames: number = 0;
+  _lastDecodedFrames: number = 0;
+  _lastTime: number = 0;
+  _mediaSourceAdapter: IMediaSourceAdapter;
+  _config: Object = {};
+  _videoElement: HTMLVideoElement;
+  static _logger: any = getLogger('droppedFramesWatcher');
+
+  constructor(mediaSourceAdapter: IMediaSourceAdapter, config: Object, videoElement: HTMLVideoElement) {
+    super();
+    this._mediaSourceAdapter = mediaSourceAdapter;
+    if (this._mediaSourceAdapter.hasFPSDropMechanism) {
+      return;
+    }
+    this._config = config;
+    this._videoElement = videoElement;
+    this._droppedFramesInterval = setInterval(() => this._maybeAdjustBitrateQuality(), this._config.fpsDroppedFramesInterval);
+  }
+
+  _maybeAdjustBitrateQuality(): void {
+    if (typeof this._videoElement.getVideoPlaybackQuality === 'function') {
+      const videoPlaybackQuality = this._videoElement.getVideoPlaybackQuality();
+      this._checkFPS(videoPlaybackQuality.droppedVideoFrames, videoPlaybackQuality.totalVideoFrames);
+    } else {
+      // $FlowFixMe
+      this._checkFPS(this._videoElement.webkitDroppedFrameCount, this._videoElement.webkitDecodedFrameCount);
+    }
+  }
+
+  _checkFPS(droppedFrames: number, decodedFrames: number): void {
+    try {
+      const currentTime = performance.now();
+      if (decodedFrames) {
+        if (this._lastTime) {
+          const currentPeriod = currentTime - this._lastTime,
+            currentDropped = droppedFrames - this._lastDroppedFrames,
+            currentDecoded = decodedFrames - this._lastDecodedFrames,
+            droppedFPS = (1000 * currentDropped) / currentPeriod;
+          if (droppedFPS > 0) {
+            DroppedFramesWatcher._logger.debug('checkFPS : droppedFPS/decodedFPS:' + droppedFPS / ((1000 * currentDecoded) / currentPeriod));
+            if (currentDropped > this._config.fpsDroppedMonitoringThreshold * currentDecoded) {
+              this._mediaSourceAdapter.restrictBandwidth();
```

Comment:
`restrictBandwidth` is very coupled to specific functionality of what you are trying to achieve in this issue, but in general what you want to have is ABR control API - so `setMaxBitrate` is more suitable and more generic.
You will however need to save a reference to current track which you can get via the event handler fro the adapter